modules/indexing/src/main/java/org/apache/ignite/internal/processors/query/stat/ColumnStatisticsCollector.java

```diff
@@ -0,0 +1,224 @@
+import java.nio.ByteBuffer;
+import java.nio.charset.StandardCharsets;
+import java.util.Comparator;
+import java.util.List;
+
+import org.apache.ignite.internal.processors.query.stat.hll.HLL;
+import org.gridgain.internal.h2.table.Column;
+import org.gridgain.internal.h2.value.Value;
+
+/**
+ * Collector to compute statistic by single column.
+ */
+public class ColumnStatisticsCollector {
+
+    /** */
+    private final Column col;
+
+    /** Hyper Log Log structure */
+    private final HLL hll = buildHll();
+
+    /** Minimum value. */
+    private Value min = null;
+
+    /** Maximum value. */
+    private Value max = null;
+
+    /** Total vals in column. */
+    private long total = 0;
+
+    /** Total size of all non nulls values (in bytes).*/
+    private long size = 0;
+
+    /** Temporary byte buffer just to avoid unnecessary object creation. */
+    private ByteBuffer bb;
+
+    /** Column value comparator. */
+    private final Comparator<Value> comp;
+
+    /** Null values counter. */
+    private long nullsCnt;
+
+    /** Hasher. */
+    private final Hasher hash = new Hasher();
+
+    /**
+     * Constructor.
+     *
+     * @param col column to collect statistics by.
+     * @param comp column values comparator.
+     */
+    public ColumnStatisticsCollector(Column col, Comparator<Value> comp) {
+        this.col = col;
+        this.comp = comp;
+    }
+
+    /**
+     * Try to fix unexpected behaviour of base Value class.
+     *
+     * @param value value to convert
+     * @return byte array
+     */
+    private byte[] getBytes(Value value) {
+        switch (value.getValueType()) {
+            case Value.STRING:
+                String strValue = value.getString();
+                return strValue.getBytes(StandardCharsets.UTF_8);
+            case Value.BOOLEAN:
+                return value.getBoolean() ? new byte[]{1} : new byte[]{0};
+            case Value.DECIMAL:
+            case Value.DOUBLE:
+            case Value.FLOAT:
+                return value.getBigDecimal().unscaledValue().toByteArray();
+            case Value.TIME:
+                return BigInteger.valueOf(value.getTime().getTime()).toByteArray();
+            case Value.DATE:
+                return BigInteger.valueOf(value.getDate().getTime()).toByteArray();
+            case Value.TIMESTAMP:
+                return BigInteger.valueOf(value.getTimestamp().getTime()).toByteArray();
+            default:
+                return value.getBytes();
+        }
+    }
+
+    public void add(Value val) {
+        total++;
+
+        if (isNull(val)) {
+            nullsCnt++;
+
+            return;
+        }
+
+        byte bytes[] = getBytes(val);
+        size += bytes.length;
+
+        hll.addRaw(hash.fastHash(bytes));
+
+        if (null == min || comp.compare(val, min) < 0)
+            min = val;
+
+        if (null == max || comp.compare(val, max) > 0)
+            max = val;
+    }
+
+    private boolean isNull(Value v) {
```

Comment:
Javadoc.
Let's make a static method in some utility class for this? e.g. H2Utils?